### ISSUE TITLE
I found Objectness in ComputeLoss which might need improvement

### ISSUE THREAD

**Issue by XHBrain** [ultralytics/yolov5#3605]
Hi,
sorry for disturbing you. I found one part in your code which might be improvement.
There maybe several GTs match the same anchor when calculate `ComputeLoss` in the scene with dense targets.

[`# Objectness`<br/>`tobj[b, a, gj, gi] = (1.0 - self.gr) + self.gr * iou.detach().clamp(0).type(tobj.dtype)  # iou ratio`](https://github.com/ultralytics/yolov5/blob/31336075609a3fbcb4afe398eba2967b22056bfa/utils/loss.py#L136)

The `tobj` is filled with the location of `[b, a, gj, gi]` which will be covered with the next same location. 
The final tobj is the bottom right GT in `build_targets`.
It may be not reasonable. The improvement of matching the GT with largest IOU could be used with the following modification.

source:
`tobj[b, a, gj, gi] = (1.0 - self.gr) + self.gr * iou.detach().clamp(0).type(tobj.dtype)  # iou ratio`
should be replaced with:
```
score_iou = iou.detach().clamp(0).type(tobj.dtype)
sort_id = torch.argsort(score_iou)
b, a, gj, gi, score_iou = b[sort_id], a[sort_id], gj[sort_id], gi[sort_id], score_iou[sort_id]
tobj[b, a, gj, gi] = (1.0 - self.gr) + self.gr * score_iou  # iou ratio
```

I just tried in the private dataset with the scene with very dense objects, the performance will be improved with above change.

**Comment by glenn-jocher** [comment]
@XHBrain hi, thanks for the loss function idea! The best way to test your change would be to train two models on a public dataset like VOC from scratch, one with your proposed change, one with the default code:
```python
python train.py --batch 64 --weights '' --cfg yolov5s.yaml --data voc.yaml --epochs 300 --cache --img 512 --hyp hyp.finetune.yaml --project VOC
```

If you could do this and report back both results that would be great!

**Comment by glenn-jocher** [comment]
@XHBrain also can you submit a PR please with your changes? Then we can checkout the branch and test the change ourselves, and if it works then we can merge into master for everyone to use. Thanks!

**Comment by XHBrain** [comment]
> @XHBrain also can you submit a PR please with your changes? Then we can checkout the branch and test the change ourselves, and if it works then we can merge into master for everyone to use. Thanks!

Hi @glenn-jocher, Thanks for your response. I commited the change for code review, please check it. 
https://github.com/XHBrain/yolov5/tree/max_iou_objectLoss
You could choose dataset with dense objects for this change verfication.

**Comment by glenn-jocher** [comment]
@XHBrain thanks! I've created PR #3610 from your branch, I'll check it out and experiment on it a bit.

**Comment by daikankan** [comment]
@XHBrain your change resort the index, but nothing changes, how can you improve the results?  I'm a confused, am I misunderstood ?

**Comment by glenn-jocher** [comment]
@XHBrain I have an IoU-sort experiment ongoing at https://wandb.ai/glenn-jocher/study-mixup-coco

Was underperforming around epoch 50 but around epoch 80 seems to be slightly overperforming the baseline. This is running on EC2 spot instance though which unfortunately forces restarts and resumes too often, so I don't know if the results are reliable.

**Comment by XHBrain** [comment]
Hi @glenn-jocher, so many thanks for your verification. From your training result, I see that my change almost has no benifit for that dataset. It was curious that the change worked on my dense dataset. Anyway, I will do more experiments and analysis on the change. Thanks again.

**Comment by glenn-jocher** [comment]
@XHBrain unfortunately the training has been restarted so many times due to spot restarts that the training results are no longer reliable. I'm going to need to secure some persistent hardware to rerun the tests. This is a bit of a problem for us at the moment...

**Comment by 791136190** [comment]
I also found this problem, and at the same time I think we should add the de-duplication operation of box regression.

**Comment by glenn-jocher** [comment]
@791136190 yes, right now one anchor fits all matching targets for cls and box, but objectness losses are only applied once per anchor. This is a bit odd on the objectness loss, but seems to work well currently.

EDIT: If you have any other ideas please let us know, or even better try them out so we have some quantitative before and after numbers we can compare.

**Comment by isJunCheng** [comment]
I used the sorting function in a custom dataset and the result has a decrease instead. The anchor with multiple sample matches should have higher iou, not be covered by low iou. **The theory is correct, but the experimental result is opposite** and I am puzzled. Can someone help me with this please?

**Comment by glenn-jocher** [comment]
@isJunCheng this could depend on the specifics of your dataset, the model architecture, and training hyperparameters. Your observations are insightful and important. It seems like an issue that warrants further investigation and perhaps a deeper understanding of the trade-offs involved. We greatly appreciate your input.
